### DISCUSSION TITLE
Is it possible to persist cache query?

### DISCUSSION THREAD

**Discussion by Jekiwijaya** [TanStack/query#92] [General] [answered]

**Comment by deevus** [comment]
I assume you mean you want to persist the query cache? Looking at the code, the cache is just an array called `queries`. It would be nice to be able to override this and use `localStorage` for example

**Comment by tannerlinsley** [comment]
You're more than welcome to do try that out, but as of today, it's not a first class citizen mostly because invalidation would need to be set up to essentially clear out old persisted queries and that's something that would be difficult to make great for a majority of use cases.

**Comment by deevus** [comment]
Would you accept a PR that decouples the cache implementation? The default implementation would function the same as it does now.

A `localStorage` implementation could be provided for example by a `react-query-cache-local-storage` package.

**Comment by tannerlinsley** [comment]
I would be open to discussing a proposal

**Comment by BrandonSmith** [comment]
@deevus Did you make progress on persisting and seeding the cache? I'm interested in using react-query in a React Native app and with the app lifecycle, seeding the cache between app restarts is an important consideration.

**Comment by tannerlinsley** [comment, marked as answer]
This is really naive, but you could try this:

```js
import { queryCache } from 'react-query'

// Persist to wherever using the super-secret `queries` object

const queriesWithData = Object.values(queryCache.queries).map(query => ({
  queryKey: query.queryKey,
  data: query.state.data
}))

localStorage.setItem('queries', JSON.stringify(queriesWithData))


// Hydrate from localStorage

const queriesWithData = JSON.parse(localStorage.getItem('queries'))

queriesWithData.forEach(query => {
  queryCache.setQueryData(query.queryHash, query.data)
})
```

This is highly experimental and undocumented, so you can be my guinea pig :)

**Comment by deevus** [comment]
I got busy but would still love this. 

@tannerlinsley that looks very interesting :)

**Comment by BrandonSmith** [comment]
@tannerlinsley And use the `queryCache.subscribe` to get notified when the cache changes?

**Comment by tannerlinsley** [comment]
Yeah!

**Comment by jgoux** [comment]
I'm also very interested by an offline-first approach! I'll try to be a guinea pig and experiment with this! 😄

**Comment by cliffordp** [comment]
https://github.com/LuudJanssen/localforage-cache worked a treat, especially setting its key like this:

```
import { stableStringify, useQuery } from 'react-query';

/**
 * Consistently get the queryKey to use with React Query, optionally casting to string for use as a local storage key.
 *
 * @param {String} endpoint
 * @param {Object} params
 * @param {Boolean} toSerializedString For use with setting local storage.
 *
 * @returns {[{String}, {Object}]}|{String}
 */
const getQueryKey = (
	endpoint,
	params,
	toSerializedString = false,
) => {
	let arr = [
		endpoint,
		params,
	];

	// https://github.com/tannerlinsley/react-query/blob/v1.2.1/src/config.js#L59
	if ( toSerializedString ) {
		arr = stableStringify( arr );
	}

	return arr;
};


/**
 * Fetch data from the API or, if available and not expired, from local storage cache.
 *
 * @param {String} endpoint
 * @param {Object} params
 */
const fetcher = async ( endpoint, params ) => {
	console.log( 'fetcher' );

	//
	// The `true` is the special bit!
	//
	const queryKeyString = getQueryKey( endpoint, params, true );

	// todo: way to force fresh API hit
	//apiCache.removeItem( queryKeyString );

	const cacheData = await apiCache.getItem( queryKeyString );

	if ( null !== cacheData ) {
		console.log( 'Got API data from local cache.' );
		return cacheData;
	}

	const url = getApiUrl( endpoint, params );
	console.log( url );

	try {
		const data = await fetch(...);
		console.log( 'API was hit: ' + endpoint );
		console.log( data );

		await apiCache.setItem( queryKeyString, data );
		console.log( 'Set API data to local cache.' );

		return data;
	} catch ( err ) {
		console.log( err );
		throw err;
	}
};

```

**Comment by cdaringe** [comment]
i think we should make this official to hydrate the cache-post-SSR. @tannerlinsley--is there a different means to hydrate post-SSR?

**Comment by 0xPetra** [comment]
Thanks! Here I share the hook I created. The implementation worked out for me in React Native and is applicable for React, just need to change AsyncStorage for whatever you use as localStorage:

Calling Read hook in UI component:
  ```js
const {data} = useReadItem(item);
```

Read hook
```js
export const useReadItem = (item?: string) => {
  const queryKey = ['QUERY-STRING', String(item)];

  return useQuery(queryKey, restRequest, {
    enabled: address,
    refetchOnWindowFocus: true,
    onSuccess: (data) => writeToStorage(queryKey, data),
  });
};
```

Manage storage Hook
```js
import {queryCache} from 'react-query';
import AsyncStorage from '@react-native-community/async-storage';

// Persist to wherever using the super-secret object
const writeToStorage = async (queryKey: string, data: any) => {
  let storageData = await AsyncStorage.getItem('queries');

  storageData = {
    ...JSON.parse(storageData ?? '{}'),
    [queryKey]: data,
  };
  AsyncStorage.setItem('queries', JSON.stringify(storageData));
};

// Hydrate from localStorage

const readFromStorage = async () => {
  const storageData = await AsyncStorage.getItem('queries');

  if (storageData !== null) {
    const queriesWithData = JSON.parse(storageData);

    for (const queryKey in queriesWithData) {
      const data = queriesWithData[queryKey];
      const queryKeyParsed = JSON.parse(queryKey);
      queryCache.setQueryData(queryKeyParsed, data);
    }
  }
};

export {readFromStorage, writeToStorage};

```

What I liked about this approach is that you can choose which queries to store and you only access local storage once when the app is accesed for the first time using a useEffect and calling `readFromStorage`.

**Reply by inoumen** [reply]
It does not work for v3

**Comment by JeremyBradshaw7** [comment]
Yes we'd need something like this in a React Native project to persist over pause/resume presumably.

**Comment by cliffordfajardo** [comment]
React query v3 has an experimental support for persisting data to localstorage using "persist-localstorage" 😄 
- https://react-query.tanstack.com/plugins/persist-localstorage#_top

**Reply by Assem-Hafez** [reply]
looks promising but still needs whitelisting/blacklisting by `querykey` or `queryhash`

**Reply by harika-sabbella** [reply]
@cliffordfajardo I tried using this in a React Native project and I get an error immediately `ReferenceError: Can't find variable: localStorage` -- looks like it's happening on line 31 in `lib/persist-localstorage-experimental/index.js`.

**Reply by cliffordfajardo** [reply]
@harika-sabbella  - localStorage unfortunately only exists inside of web browsers. It don't exist in the environment where `react-native` runs since the code that eventually gets run on the user device is not a browser. 

 Would be a good idea to open up an issue about this. The equivalent of `localStorage` that exists in `react-native` is called [AsyncStorage](https://reactnative.dev/docs/asyncstorage). `react-query` likely needs a new plugin created by the community like the one for localStorage

**Reply by harika-sabbella** [reply]
I was actually wondering if this same code could maybe be leveraged to optionally pass in a local storage type so that it can work for both mobile (mobile would use AsyncStorage) and web (would continue to use localStorage)?

**Reply by ivanjeremic** [reply]
any demo on how to set the key, I cannot find any `set` or `get` function?

**Reply by harika-sabbella** [reply]
Are you wondering about how to set/get things from `AsyncStorage`? If so, here is the documentation for it https://reactnative.dev/docs/asyncstorage#getitem and https://reactnative.dev/docs/asyncstorage#setitem

**Reply by inoumen** [reply]
Are there any workarounds to whitelist/blacklist queries?

**Reply by a-eid** [reply]
@ivanjeremic `AsyncStorage` has almost the same api as `localstorage` except that it's methods are async.

**Reply by kokoskiwi** [reply]
when will that feature be stable?

**Reply by TkDodo** [reply]
Likely with v4.

**Comment by a-eid** [comment]
@tannerlinsley how efficient is it to persist the cache ?

**Reply by bombillazo** [reply]
I'm assuming it'll depend on your local storage. If you are using React Native I recommend [this storage](https://github.com/ammarahm-ed/react-native-mmkv-storage).

**Comment by muxcortoi** [comment]
Does createWebStoragePersistor-experimental uses "cacheTime" QueryClient property? Because I persist the cache but its forever in localStorage.

**Reply by TkDodo** [reply]
Yes, it’s documented here: https://react-query.tanstack.com/plugins/persistQueryClient#usage

**Comment by kinthaiofficial** [comment]
Persisting TanStack Query cache is now well-supported — here's the current recommended approach:

**Using `@tanstack/query-persist-client-core` (official):**
```typescript
import { createSyncStoragePersister } from '@tanstack/query-sync-storage-persister';
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client';

const persister = createSyncStoragePersister({
  storage: window.localStorage,
});

function App() {
  return (
    <PersistQueryClientProvider
      client={queryClient}
      persistOptions={{
        persister,
        maxAge: 1000 * 60 * 60 * 24,  // 24 hours
        buster: 'v1',  // increment to invalidate old cache on deploy
      }}
    >
      <YourApp />
    </PersistQueryClientProvider>
  );
}
```

**AsyncStorage for React Native:**
```typescript
import { createAsyncStoragePersister } from '@tanstack/query-async-storage-persister';
import AsyncStorage from '@react-native-async-storage/async-storage';

const persister = createAsyncStoragePersister({
  storage: AsyncStorage,
});
```

**Key considerations:**

**Cache size**: LocalStorage has a 5MB limit. Large query results will fail silently if they exceed it. Serialize only the queries you actually need persisted using `queryClient.setQueryDefaults` or filtering in `persistOptions.dehydrateOptions`.

**Stale data on restore**: Persisted data is restored but marked stale, triggering a background refetch. The user sees immediate data (from cache) while fresh data loads — this is the "stale-while-revalidate" pattern.

**Sensitive data**: Don't persist queries containing auth tokens, PII, or other sensitive data. Filter them out in `dehydrateOptions`.

**For agent-driven apps**: Persisting the query cache means an agent can pick up a session where it left off — the data context is immediately available without re-fetching.

**Comment by kinthaiofficial** [comment]
Persisting the React Query cache is well-supported with the `@tanstack/query-sync-storage-persister` or `@tanstack/query-async-storage-persister` packages:

```typescript
import { QueryClient } from '@tanstack/react-query';
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client';
import { createSyncStoragePersister } from '@tanstack/query-sync-storage-persister';
import { createAsyncStoragePersister } from '@tanstack/query-async-storage-persister';

// Option 1: localStorage (synchronous, limited size ~5MB)
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 1000 * 60 * 5,  // 5 minutes
      gcTime: 1000 * 60 * 60 * 24,  // 24 hours (old: cacheTime)
    },
  },
});

const persister = createSyncStoragePersister({
  storage: window.localStorage,
  key: 'REACT_QUERY_CACHE',
  throttleTime: 1000,  // Debounce writes
});

// Option 2: IndexedDB (async, much larger storage)
import { get, set, del } from 'idb-keyval';

const asyncPersister = createAsyncStoragePersister({
  storage: {
    getItem: (key) => get(key),
    setItem: (key, value) => set(key, value),
    removeItem: (key) => del(key),
  },
});

// Wrap your app
function App() {
  return (
    <PersistQueryClientProvider
      client={queryClient}
      persistOptions={{
        persister,
        maxAge: 1000 * 60 * 60 * 24,  // 24 hours
        buster: process.env.DEPLOY_VERSION,  // Cache busting on deploy
      }}
    >
      <YourApp />
    </PersistQueryClientProvider>
  );
}
```

**Key considerations**:
- `buster` string forces cache invalidation on deploys — use your build hash
- Use `gcTime` (v5) or `cacheTime` (v4) longer than `staleTime` so data persists between sessions
- Tag queries you don't want persisted: `queryKey: ['temp-data'], meta: { persist: false }`
- Filter queries with `dehydrateOptions`: only persist specific query keys

For agent applications where you want to cache the agent's knowledge state client-side, persistence is particularly useful: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture
